### PR TITLE
Shooter

### PR DESCRIPTION
**Pull request by vichik123**
Had to rename the branch and the previous PR was deleted.

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -1,50 +1,33 @@
...
-        public static final double SENSOR_TOLERANCE = 0.2; // Sensor tolerance for state space.
+        public static final double MODEL_TOLERANCE = 10; // Model tolerance for state space.
+        public static final double SENSOR_TOLERANCE = 0.1; // Sensor tolerance for state space.
+        public static final double VELOCITY_TOLERANCE = 0.15; // Velocity tolerance for state space.
+        public static final double COST_LQR = 55; // Cost lqr for state space.
 
-        public static final double J = 0.00032; // Moment of inertia for state space. [kg*m^2]
+        public static final double J = 0.00218; // Moment of inertia for state space. [kg*m^2]
         public static final double GEAR_RATIO = 1; // Gear ratio for encoder (placeholder).
+        public static final double NEUTRAL_DEADBAND = 0.1;
```
Comment (barel13) [inline src/main/java/frc/robot/Constants.java]:
```suggestion
        public static final double NEUTRAL_DEADBAND = 0.1; // [%]
```

File: src/main/java/frc/robot/Ports.java [hunk 2]
```diff
@@ -1,15 +1,24 @@
...
+
 public final class Ports {
+    public static class ExampleSubsystem {
+        public static final int MAIN = 0;
+        public static final int AUX = 0;
+        public static final boolean IS_MAIN_INVERTED = false;
+        public static final boolean IS_AUX_INVERTED = false;
+        public static final boolean IS_MAIN_SENSOR_INVERTED = false;
+        public static final boolean IS_AUX_SENSOR_INVERTED = false;
+    }
```
Comment (barel13) [inline src/main/java/frc/robot/Ports.java]: Why do you care about example subsystem

File: src/main/java/frc/robot/Ports.java [hunk 3]
```diff
@@ -1,15 +1,24 @@
...
+
     public static class Shooter {
-        public static final int SOLENOID = 0;
-        public static final int MAIN_MOTOR = 0;
-        public static final boolean MAIN_INVERTED = false;
-        public static final boolean MAIN_SENSOR_PHASE = false;
+        public static final int MOTOR = 0; // Main motor port.
+        public static final TalonFXConfiguration CONFIGURATION = new TalonFXConfiguration();
```
Comment (barel13) [inline src/main/java/frc/robot/Ports.java]: Should be in Constants

File: src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -1,50 +1,33 @@
...
-        public static final double Cd = 0.4; // Drag coefficient of the ball.
-        public static final double AIR_DENSITY = 1.225; // Average density of air. [kg/m^3]
 
-        public static final double NOMINAL_VOLTAGE = 12.0; // Nominal voltage. [V]
-        public static final double WHEEL_RADIUS = 1; // Radius of the flywheel. [m]
-        public static final int TICKS_PER_REVOLUTION = 2048; // Ticks per revolution of the shooter motor. [tick/c]
-        public static final double TICKS_PER_METER = Math.pow(WHEEL_RADIUS, 2) * Math.PI
-                / TICKS_PER_REVOLUTION; // Ticks per meter of the shooter flywheel. [tick/m]
+    public static class Shooter {
+        public static final double TICKS_PER_REVOLUTION = 2048; // Ticks per revolution of the shooter motor. [tick]
```
Comment (barel13) [inline src/main/java/frc/robot/Constants.java:27]: If you use this constant, the values will be in rps, you can convert it so it would be in rpm

File: src/main/java/frc/robot/commandgroups/Outtake.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import java.util.function.DoubleSupplier;
+
+public class Outtake extends ParallelCommandGroup {
+
+    public Outtake(Shooter shooter, DoubleSupplier distance) {
+        addCommands(
+                new Shoot(shooter, () -> Shooter.getSetpointVelocity(distance.getAsDouble()))
+        );
+    }
+}
```
Comment (barel13) [inline src/main/java/frc/robot/commandgroups/Outtake.java]: command groups should be in integration, not in the shooter

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 6]
```diff
@@ -12,51 +15,108 @@
...
 public class Shooter extends SubsystemBase {
-    private static final Shooter INSTANCE = new Shooter();
-    private final UnitModel unitModel = new UnitModel(TICKS_PER_METER);
-    private final WPI_TalonFX mainMotor = new WPI_TalonFX(MAIN_MOTOR);
+    private static Shooter INSTANCE;
+    private final UnitModel unitModel = new UnitModel(TICKS_PER_REVOLUTION);
+    private final WPI_TalonFX motor = new WPI_TalonFX(MOTOR);
     private final LinearSystemLoop<N1, N1, N1> linearSystemLoop;
-    private final DCMotor motor = DCMotor.getFalcon500(1);
+    private final Timer timer = new Timer();
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: You can remove the timer

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 7]
```diff
@@ -66,21 +126,59 @@ else if (stateSpaceType.equals("Inertia"))
...
-        linearSystemLoop.setNextR(VecBuilder.fill(velocity));
-        linearSystemLoop.correct(VecBuilder.fill(getVelocity()));
-        linearSystemLoop.predict(timeInterval);
+    /**
+     * Sets the velocity of the motor.
+     *
+     * @param velocity is the velocity setpoint. [rpm]
+     */
+    public void setVelocity(double velocity) {
+        linearSystemLoop.setNextR(VecBuilder.fill(Units.rotationsToRadians(Utils.secondsToMinutes(velocity))));
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: the inputted velocity is in rpm, so:
the velocity is converted from seconds to minutes (same as minutes to hours): the units are [rph] -> rotations to radians: units: radians per hour...

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 8]
```diff
@@ -66,21 +126,59 @@ else if (stateSpaceType.equals("Inertia"))
...
+            SmartDashboard.putNumber("ve", velocity);
+            flywheelSim.setInputVoltage(MathUtil.clamp(linearSystemLoop.getU(0), -NOMINAL_VOLTAGE, NOMINAL_VOLTAGE));
+            simCollection.setIntegratedSensorVelocity(unitModel.toTicks100ms(Utils.secondsToMinutes(flywheelSim.getAngularVelocityRPM())));
+        } else {
+            motor.setVoltage(MathUtil.clamp(linearSystemLoop.getU(0), -NOMINAL_VOLTAGE, NOMINAL_VOLTAGE));
+        }
+    }
 
-        mainMotor.set(ControlMode.PercentOutput, linearSystemLoop.getU(0) / NOMINAL_VOLTAGE);
+    public void setPower(double output) {
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: documentation

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 9]
```diff
@@ -66,21 +126,59 @@ else if (stateSpaceType.equals("Inertia"))
...
+        currentTime = Timer.getFPGATimestamp();
+    }
+
+    @Override
+    public void simulationPeriodic() {
+        lastTime = currentTime;
+        SmartDashboard.putNumber("velocity", flywheelSim.getAngularVelocityRPM());
+        SmartDashboard.putNumber("power", simCollection.getMotorOutputLeadVoltage());
+        flywheelSim.update(LOOP_PERIOD);
+        currentTime = Timer.getFPGATimestamp();
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:172]: please group currentTime with lastTime

File: src/main/java/frc/robot/subsystems/shooter/commands/Shoot.java [hunk 10]
```diff
@@ -1,15 +1,33 @@
...
+
+public class Shoot extends CommandBase {
+    private final Shooter shooter;
+    private final DoubleSupplier velocitySetpoint;
+
+    /**
+     * Constructor.
+     *
+     * @param shooter          is the shooter subsystem.
+     * @param velocitySetpoint is the setpoint for the shooter wheel. [rps]
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/commands/Shoot.java]: rpm

File: src/main/java/frc/robot/subsystems/shooter/commands/Shoot.java [hunk 11]
```diff
@@ -1,15 +1,33 @@
...
+     */
+    public Shoot(Shooter shooter, DoubleSupplier velocitySetpoint) {
+        this.shooter = shooter;
+        this.velocitySetpoint = velocitySetpoint;
+        addRequirements(shooter);
+    }
+
+    @Override
+    public void execute() {
+        shooter.setVelocity(Shooter.getSetpointVelocity(velocitySetpoint.getAsDouble()));
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/commands/Shoot.java]: You don't need to use the function `getSetpointVelocity`, it should be passed in the supplier (look at Outtake for example)

File: src/main/java/frc/robot/subsystems/shooter/commands/bits/ShootAtMaximalVelocity.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
+package frc.robot.subsystems.shooter.commands.bits;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.shooter.Shooter;
+
+public class ShootAtMaximalVelocity extends CommandBase {
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/commands/bits/ShootAtMaximalVelocity.java]: Why do you need this command?

File: src/main/java/frc/robot/utils/Utils.java [hunk 13]
```diff
@@ -26,4 +26,23 @@ public static double rotationalDeadband(double input, double threshold) {
...
+
+    /**
+     * Converts seconds into minutes.
+     * Used for converting rpm into rps in shooter branch.
+     * @param seconds is the input of seconds. [s]
+     * @return how many minutes the seconds amount to. [min]
+     */
+    public static double secondsToMinutes(double seconds){
+        return seconds / 60.0;
+    }
```
Comment (barel13) [inline src/main/java/frc/robot/utils/Utils.java]: (I'm sorry) Ctrl+Alt+L

**Review by barel13 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/utils/Utils.java (hunk 13 above)
Comment (katzuv) [reply src/main/java/frc/robot/utils/Utils.java]: Proud of you ❤️

File: README.md [hunk 14]
```diff
@@ -1,4 +1,4 @@
-robot-2022
+robot 2022 - [Stardust](https://www.merriam-webster.com/dictionary/stardust)
```
Comment (katzuv) [inline README.md]: Please, this change doesn't belong in this branch. Move it to another branch.
You have to pay attention and make sure each branch contains only the relevant changes. (@adamv17)

File: src/main/java/frc/robot/Ports.java (hunk 2 above)
Comment (katzuv) [reply src/main/java/frc/robot/Ports.java]: Also, this change shouldn't be made in this branch.

File: src/main/java/frc/robot/Ports.java (hunk 3 above)
Comment (katzuv) [reply src/main/java/frc/robot/Ports.java]: Why? Configurations are usually stored in the ports file.

File: src/main/java/frc/robot/RobotContainer.java [hunk 15]
```diff
@@ -4,19 +4,20 @@
...
-import frc.robot.subsystems.hood.commands.HoodDefaultCommand;
 import frc.robot.subsystems.shooter.commands.Shoot;
 import frc.robot.valuetuner.ValueTuner;
 import webapp.Webserver;
 
+import static frc.robot.Constants.Control.RIGHT_TRIGGER_DEADBAND;
+
 public class RobotContainer {
     // The robot's subsystems and commands are defined here...
+    public ExampleSubsystem exampleSubsystem = ExampleSubsystem.getInstance();
```
Comment (katzuv) [inline src/main/java/frc/robot/RobotContainer.java]: Why? And why here?

File: src/main/java/frc/robot/RobotContainer.java [hunk 16]
```diff
@@ -4,19 +4,20 @@
...
     private final XboxController xbox = new XboxController(Ports.Controls.XBOX);
     private final JoystickButton a = new JoystickButton(xbox, XboxController.Button.kA.value);
-    private final Trigger rt = new Trigger(() -> xbox.getRightTriggerAxis() > 0);
-    private final Hood hood = Hood.getINSTANCE();
```
Comment (katzuv) [inline src/main/java/frc/robot/RobotContainer.java]: Why here?

File: src/main/java/frc/robot/RobotContainer.java [hunk 17]
```diff
@@ -35,15 +36,12 @@ public RobotContainer() {
     }
 
     private void configureDefaultCommands() {
-        hood.setDefaultCommand(new HoodDefaultCommand(a::get));
+//        shooter.setDefaultCommand(new TestPistonPressure(xbox::getLeftY, shooter));
```
Comment (katzuv) [inline src/main/java/frc/robot/RobotContainer.java]: Don't commit commented-out code.

File: src/main/java/frc/robot/RobotContainer.java [hunk 18]
```diff
@@ -35,15 +36,12 @@ public RobotContainer() {
...
+        shooter.setDefaultCommand(new Shoot(shooter, () -> 2));
     }
 
     private void configureButtonBindings() {
-        /*
-        Currently, the shooting is at 20 meters per second. This will be changed once
-        the vision becomes available for use.
-         */
-        rt.whileActiveOnce(new Shoot(20));
+//        rightTrigger.whileActiveContinuous(new Outtake(shooter, () -> 4));
```
Comment (katzuv) [inline src/main/java/frc/robot/RobotContainer.java]: Don't commit commented-out code.

File: src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+import frc.robot.Ports;
+
+public class ExampleSubsystem extends SubsystemBase {
+    private static ExampleSubsystem INSTANCE = null;
+    // create fields here
+
+    private final WPI_TalonFX main = new WPI_TalonFX(Ports.ExampleSubsystem.MAIN);
+    private final WPI_TalonSRX aux = new WPI_TalonSRX(Ports.ExampleSubsystem.AUX);
+
+    private ExampleSubsystem() {
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/example/ExampleSubsystem.java]: Why are these changes needed, and why in this branch?

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 20]
```diff
@@ -12,51 +15,108 @@
...
+        linearSystemLoop = configStateSpace(true);
+        if (Robot.isSimulation()) {
+            timer.start();
+            timer.reset();
+            simCollection = motor.getSimCollection();
+        }
     }
 
+    private void configureMotor(){
+        CONFIGURATION.neutralDeadband = NEUTRAL_DEADBAND;
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: There is no advantage in creating `NEUTRAL_DEADBAND` and then using it here. Just call it in the ports file.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 21]
```diff
@@ -12,51 +15,108 @@
...
+    /**
+     * Calculates the velocity setpoint according to the distance from the target.
+     * Once the data from the shooter is acquired this function will be changed.
+     *
+     * @param distance is the distance from the target. [m]
+     * @return 15. [rpm]
+     */
+    public static double getSetpointVelocity(double distance) {
+//        return 15 * distance;
+        return 100;
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: The return value documentation doesn't match the code.
But anyway, I think this method should be in the relevant command, not here.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 8 above)
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: A method named `setPower` should receive a parameter named `power`.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 22]
```diff
@@ -66,21 +126,59 @@ else if (stateSpaceType.equals("Inertia"))
...
+     * Sets the velocity of the motor.
+     *
+     * @param velocity is the velocity setpoint. [rpm]
+     */
+    public void setVelocity(double velocity) {
+        linearSystemLoop.setNextR(VecBuilder.fill(Units.rotationsToRadians(Utils.secondsToMinutes(velocity))));
+        linearSystemLoop.correct(VecBuilder.fill(Units.rotationsToRadians(Utils.secondsToMinutes(getVelocity()))));
+        linearSystemLoop.predict(currentTime - lastTime);
+        if (Robot.isSimulation()) {
+            SmartDashboard.putNumber("ve", velocity);
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:145]: 1. Please use Shuffleboard and subtables instead of Smart Dashboard.
2. "ve" is very short and not clear. Rename it.

File: src/main/java/frc/robot/utils/Utils.java [hunk 23]
```diff
@@ -26,4 +26,23 @@ public static double rotationalDeadband(double input, double threshold) {
         return (input - (Math.signum(input) * threshold)) / (1 - threshold);
     }
 
+    /**
+     * Converts minutes into seconds.
+     * @param minutes is the input of minutes. [min]
+     * @return how many seconds are in that minutes. [s]
```
Comment (katzuv) [inline src/main/java/frc/robot/utils/Utils.java]: Maybe "these", "that minutes" doesn't sound natural to me.

File: src/main/java/frc/robot/utils/Utils.java [hunk 24]
```diff
@@ -26,4 +26,23 @@ public static double rotationalDeadband(double input, double threshold) {
         return (input - (Math.signum(input) * threshold)) / (1 - threshold);
     }
 
+    /**
+     * Converts minutes into seconds.
+     * @param minutes is the input of minutes. [min]
```
Comment (katzuv) [inline src/main/java/frc/robot/utils/Utils.java]:
```suggestion
     * @param minutes is the input of minutes [min]
```

File: vendordeps/Phoenix.json [hunk 25]
```diff
@@ -2,11 +2,12 @@
     "fileName": "Phoenix.json",
     "name": "CTRE-Phoenix",
     "version": "5.20.2",
+    "frcYear": 2022,
     "uuid": "ab676553-b602-441f-a38d-f1296eff6537",
     "mavenUrls": [
-        "https://devsite.ctr-electronics.com/maven/release/"
+        "https://maven.ctr-electronics.com/release/"
```
Comment (katzuv) [inline vendordeps/Phoenix.json]: This should be made in another branch.

File: vendordeps/WPILibNewCommands.json [hunk 26]
```diff
@@ -1,7 +1,7 @@
 {
   "fileName": "WPILibNewCommands.json",
   "name": "WPILib-New-Commands",
-  "version": "2020.0.0",
+  "version": "2022.1.1",
```
Comment (katzuv) [inline vendordeps/WPILibNewCommands.json]: This change should be made in another branch.

**Review by katzuv — CHANGES_REQUESTED** [review]

File: README.md (hunk 14 above)
Comment (adamv17) [reply README.md]: I don't why this is here, I think maybe because of an earlier merge to prepare this branch for merging

Comment (adamv17) [reply README.md]: I strongly suggest to tone down the criticism as it is not appropriate for this situation

Comment (katzuv) [reply README.md]: Sorry for the harsh comment, I didn't mean it. I do not criticize you, it's really for the benefit of the project and your wellbeing during future testing 🙂.

File: src/main/java/frc/robot/Ports.java (hunk 2 above)
Comment (vichik123) [reply src/main/java/frc/robot/Ports.java]: This was there outside of my branch. At one point I deleted it, so then I just rewrote it.

File: src/main/java/frc/robot/Constants.java (hunk 4 above)
Comment (vichik123) [reply src/main/java/frc/robot/Constants.java:27]: Could you rephrase that? I don't understand what you're trying to say.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 7 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Once again: The units are not in minutes, they are in rotation/minute. So if we substitute minute = second * 60, all in all we divide by 60. Hence seconds to minutes. Should I rename the function to rpmToRps?

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 9 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java:172]: Wdym?

File: src/main/java/frc/robot/subsystems/shooter/commands/bits/ShootAtMaximalVelocity.java (hunk 12 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/commands/bits/ShootAtMaximalVelocity.java]: 1. To check whether the solenoid stays at the correct angle with maximal velocity.
2. To check the maximal distance that the shooter can reach.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 20 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: CONFIGURATION is a static constant. Intellij doesn't allow me to call it in the ports file.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 21 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: In the future this function will return the interpolation of the setpoint velocity.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 22 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java:145]: The shuffleboard sees the smart dashboard. In addition, this is when the robot is in simulation, meaning that the smart dashboard is what the simulate Java program sees.

File: src/main/java/frc/robot/utils/Utils.java (hunk 24 above)
Comment (vichik123) [reply src/main/java/frc/robot/utils/Utils.java]: Adam taught us documentation a little bit differently. @adamv17 could you take a look?

File: src/main/java/frc/robot/Ports.java (hunk 3 above)
Comment (barel13) [reply src/main/java/frc/robot/Ports.java]: Because the neutral deadband doesn't related to the ports...

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 7 above)
Comment (barel13) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: You should definitely rename the function.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 9 above)
Comment (barel13) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java:172]: Just move`currentTime = ...` to be right after `lastTime`

File: src/main/java/frc/robot/subsystems/shooter/commands/bits/ShootAtMaximalVelocity.java (hunk 12 above)
Comment (barel13) [reply src/main/java/frc/robot/subsystems/shooter/commands/bits/ShootAtMaximalVelocity.java]: You don't really need this function, if I would tell the shooter to shoot at 2,000,000 rpm I don't think he would shoot at this rate...

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 22 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java:145]: > The shuffleboard sees the smart dashboard

I see. Although it is still more organized that way, especially when you (will shortly) have multiple subsystems.
> the smart dashboard is what the simulated Java program sees.

WDYM?

File: src/main/java/frc/robot/utils/Utils.java (hunk 24 above)
Comment (katzuv) [reply src/main/java/frc/robot/utils/Utils.java]: OK, that's fine if you have your own convention. But just FYI, the [Javadoc article on Oracle's website](https://www.oracle.com/il-en/technical-resources/articles/java/javadoc-tool.html#format) states otherwise.

File: src/main/java/frc/robot/Ports.java (hunk 3 above)
Comment (katzuv) [reply src/main/java/frc/robot/Ports.java]: But it _is_ a configuration of the Talon FX.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 20 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: You can do something like that in the ports file (@adamv17, would like your opinion):
```java
public static TalonFXConfiguration getMotorConfiguration() {
        final TalonFXConfiguration configuration = new TalonFXConfiguration();
        configuration.neutralDeadband = 0.1;
        // More configurations...
        return configuration;
        }
```
Or, perhaps better:
```java
public static WPI_TalonFX getMotor() {
        final WPI_TalonFX motor = new WPI_TalonFX(0);
         motor.configNeutralDeadband(0.1);
        // More configurations...
        return motor;
    }
```
This way you don't even need the configuration object! 🆒
Also:
> Intellij doesn't allow me to call it in the ports file.

It's not IntelliJ IDEA limiting you, it's Java itself 🙂.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 21 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: @littlebro123, what's you thought about that?
> But anyway, I think this method should be in the relevant command, not here.

File: src/main/java/frc/robot/subsystems/shooter/commands/bits/ShootAtMaximalVelocity.java (hunk 12 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/commands/bits/ShootAtMaximalVelocity.java]: > To check whether the solenoid stays at the correct angle with maximal velocity.

WDYM?
> To check the maximal distance that the shooter can reach.

You can just use the `Shoot` command you wrote with a high, constant velocity setpoint.

**Review by katzuv — CHANGES_REQUESTED** [review]
Just a few additional comments 🙂.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 22 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java:145]: > WDYM?

I mean that the simulateJava gradle task sees the SmartDashboard when in the simulation.

File: src/main/java/frc/robot/utils/Utils.java (hunk 24 above)
Comment (vichik123) [reply src/main/java/frc/robot/utils/Utils.java]: Alright, but I'd rather stick to our convention.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 21 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java]: The function will be based off of the values that we measure with our testing. This objectively means that the interpolation belongs to the Shooter subsystem itself, and not the command.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java (hunk 22 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java:145]: Do you mean that `simulateJava` polls data from SmartDashboard?

Comment (vichik123) [reply src/main/java/frc/robot/subsystems/shooter/Shooter.java:145]: Yah

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 27]
```diff
@@ -12,51 +15,102 @@
...
+        configureMotor();
+        linearSystemLoop = configStateSpace(true);
+        if (Robot.isSimulation()) {
+            simCollection = motor.getSimCollection();
+        }
+    }
+
+    private void configureMotor() {
+        motor.configAllSettings(getConfiguration());
+        motor.setInverted(TalonFXInvertType.CounterClockwise);
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: Create a constant that stores the value of `TalonFXInvertType` (in ports class)

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 28]
```diff
@@ -66,21 +120,64 @@ else if (stateSpaceType.equals("Inertia"))
...
+     * Gets the velocity of the motor.
+     *
+     * @return the velocity of the motor. [rpm]
+     */
     public double getVelocity() {
-        return unitModel.toUnits(mainMotor.getSelectedSensorVelocity());
+        if (Robot.isSimulation()) {
+            return flywheelSim.getAngularVelocityRPM();
+        }
+        return unitModel.toVelocity(motor.getSelectedSensorVelocity());
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java]: is it in rpm?

File: src/main/java/frc/robot/subsystems/shooter/commands/Shoot.java (hunk 10 above)
Comment (barel13) [inline src/main/java/frc/robot/subsystems/shooter/commands/Shoot.java]: rpm (again)
```suggestion
     * @param velocitySetpoint is the setpoint for the shooter wheel. [rpm]
```

**Review by barel13 — CHANGES_REQUESTED** [review]

**Review by barel13 — APPROVED** [review]

**Review by adamv17 — APPROVED** [review]
